pennylane/shadows/classical_shadow.py

```diff
@@ -404,42 +403,48 @@ def qnode(x):
             bitstrings, recipes = qnode(x)
             shadow = qml.ClassicalShadow(bitstrings, recipes)
 
-        >>> [shadow.entropy(wires=wires, alpha=alpha, atol=1e-10) for alpha in [1., 2., 3.]]
+        >>> [shadow.entropy(wires=wires, alpha=alpha) for alpha in [1., 2., 3.]]
         [1.5419292874423107, 1.1537924276625828, 0.9593638767763727]
 
         """
 
+        if atol is not None:
+            warnings.warn(
+                "Trying to use atol in ClassicalShadow.entropy. This feature is deprecated and has been replaced by a more accurate version that projects the density matrix estimator to the closest valid density matrix in L2 norm. You can simply remove the atol keyword in your code.",
+                DeprecationWarning,
+            )
+
         global_snapshots = self.global_snapshots(wires=wires, snapshots=snapshots)
         rdm = median_of_means(global_snapshots, k, axis=0)
 
         # Allow for different log base
         div = np.log(base) if base else 1
 
-        # This was returning negative values in most cases, so commenting it out
-        # until we figure out the issue.
-        # if alpha == 2:
-        #     # special case of purity
-        #     res = -qml.math.log(qml.math.trace(rdm @ rdm))
-        #     return res / div
-
-        # Else
-        # Compute Eigenvalues and choose only those >>0
-        evs = qml.math.eigvalsh(rdm)
-        mask0 = qml.math.logical_not(qml.math.isclose(evs, 0, atol=atol))
-        mask1 = qml.math.where(evs > 0, True, False)
-        mask = qml.math.logical_and(mask0, mask1)
-        # Renormalize because of cropped evs
-        evs_nonzero = qml.math.gather(evs, mask)
-        evs_nonzero = evs_nonzero / qml.math.sum(evs_nonzero)
-
+        evs_nonzero = _project_density_matrix_spectrum(rdm)
         if alpha == 1:
             # Special case of von Neumann entropy
-            return -qml.math.sum(evs_nonzero * qml.math.log(evs_nonzero)) / div
+            return qml.math.entr(evs_nonzero) / div
 
         # General Renyi-alpha entropy
         return qml.math.log(qml.math.sum(evs_nonzero**alpha)) / (1.0 - alpha) / div
 
 
+def _project_density_matrix_spectrum(rdm):
+    """Project the estimator density matrix rdm with possibly negative eigenvalues onto the closest true density matrix in L2 norm"""
+    # algorithm below eq. (16) in https://arxiv.org/pdf/1106.5458.pdf
+    evs = qml.math.eigvalsh(rdm)[::-1]  # order from largest to smallest
+    d = len(rdm)
+    i = d
+    a = 0.0
+    for i in range(d - 1, -1, -1):
+        if evs[i] + a / i > 0:
```

Comment:
Is it ever the case that only the last eigenvalue pushes `a` to a non-negative value, so that this gets all the way to index 0?

Reply:
Yes, actually quite common. Basically every pure state $|\psi\rangle$ has this porperty becuase the "decomposition" is just $\rho = 1 |\psi \times \psi |$

Reply:
(i.e. only eigenvalue of $\rho$ is 1

Reply:
I feel like a weird thing happens in that case where `a` is just the final eigenvalue, and then we divide `a` by `0+1` and add it to the eigenvalue, and suddenly we have `2*lambda` instead of `lambda` as the return.

Reply:
Ah that is what you meant, yes! nice catch! The range should be going to 0, not -1

Reply:
I think we need to treat the case where `evs[i] + a / (i+1)` is never positive separately. I added this with a corresponding test